pkg/ipamd/datastore/data_store.go

```diff
@@ -52,10 +52,28 @@ const (
 	UnknownENIError = "datastore: unknown ENI"
 )
 
-// Enable temporary checkpoint upgrade logic.  If checkpoint file is
-// not found, query CRI to learn current pods and currently-allocated
-// IPs.  Will be disabled/removed after next release.
-const BackfillMissingCheckpointFromCRI = true
+// We need to know which IPs are already allocated across
+// ipamd/datastore restarts.  In vpc-cni <=1.6, we "stored" the
+// allocated IPs by querying kubelet's CRI.  Since this requires scary
+// access to CRI socket, and may race with CRI's internal logic, we
+// are transitioning away from from this to storing allocations
+// ourself in a file (similar to host-ipam CNI plugin).
+//
+// Because we don't want to require a node restart during CNI
+// upgrade/downgrade, we need an "expand/contract" style upgrade to
+// keep the two stores in sync:
+//
+// Migration phase0 (CNI 1.6): Read/write from CRI only.
+// Migration phase1 (CNI 1.7): Read from CRI.  Write to CRI+file.
+// Migration phase2 (CNI 1.8?): Read from file. Write to CRI+file.
+// Migration phase3 (hypothetical): Read/write from file only.
```

Comment:
Thanks a lot for the clear documentation. I'll update the readme as well in a follow up PR